File: src/ccycles/camera.cpp
/**
Copyright 2014-2017 Robert McNeel and Associates

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

#include "internal_types.h"

void cycles_camera_set_size(ccl::Session* session_id, unsigned int width, unsigned int height)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
        // TODO: find out if need to set full_width and full_height
        sce->camera->set_screen_size(width, height);
		// TODO: APIfy need_[device_]update
		sce->camera->need_flags_update = true;
		sce->camera->need_device_update = true;
	}
}

unsigned int cycles_camera_get_width(ccl::Session* session_id)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		return (unsigned int)sce->camera->get_full_width();
	}

	return 0;
}

unsigned int cycles_camera_get_height(ccl::Session* session_id)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		return (unsigned int)sce->camera->get_full_height();
	}

	return 0;
}

void cycles_camera_set_type(ccl::Session* session_id, camera_type type)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_camera_type((ccl::CameraType)type);
	}
}

void cycles_camera_set_panorama_type(ccl::Session* session_id, panorama_type type)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_panorama_type((ccl::PanoramaType)type);
	}
}

void cycles_camera_set_matrix(ccl::Session* session_id,
	float a, float b, float c, float d,
	float e, float f, float g, float h,
	float i, float j, float k, float l
	)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		ccl::Transform mat = ccl::make_transform(a, b, c, d, e, f, g, h, i, j, k, l);
		sce->camera->set_matrix(mat);
	}
}

void cycles_camera_compute_auto_viewplane(ccl::Session* session_id)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->compute_auto_viewplane();
	}
}

void cycles_camera_set_viewplane(ccl::Session* session_id, float left, float right, float top, float bottom)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->viewplane.left = left;
		sce->camera->viewplane.right = right;
		sce->camera->viewplane.top = top;
		sce->camera->viewplane.bottom = bottom;
	}

}

void cycles_camera_update(ccl::Session* session_id)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->update(sce);

		// 2023-05-22 David E. (RH-74901)
		// NOTE: This is a hack. I'm calling tag_modified() here
		// because otherwise the updated camera will get ignored in 
		// Scene::update_kernel_features() due to need_update() returning
		// false. It seems like camera->update(...) should not get called
		// while we're rendering with CyclesX. Instead, we should be
		// doing what the standalone app does:
		//   camera->set_matrix(matrix);
		//   camera->need_flags_update = true;
		//   camera->need_device_update = true;
		// The code above will set only what's needed and correctly tag
		// the camera that it's been modified.
		// Please see YouTrack item RH-74901.
		sce->camera->tag_modified();
	}
}

void cycles_camera_set_fov(ccl::Session* session_id, float fov)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_fov(fov);
	}
}

void cycles_camera_set_sensor_width(ccl::Session* session_id, float sensor_width)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_sensorwidth(sensor_width);
	}
}

void cycles_camera_set_sensor_height(ccl::Session* session_id, float sensor_height)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_sensorheight(sensor_height);
	}
}

void cycles_camera_set_nearclip(ccl::Session* session_id, float nearclip)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_nearclip(nearclip);
	}
}

void cycles_camera_set_farclip(ccl::Session* session_id, float farclip)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_farclip(farclip);
	}
}

void cycles_camera_set_aperturesize(ccl::Session* session_id, float aperturesize)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_aperturesize(aperturesize);
	}
}

void cycles_camera_set_aperture_ratio(ccl::Session* session_id, float aperture_ratio)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_aperture_ratio(aperture_ratio);
	}
}

void cycles_camera_set_blades(ccl::Session* session_id, unsigned int blades)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_blades(blades);
	}
}

void cycles_camera_set_bladesrotation(ccl::Session* session_id, float bladesrotation)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_bladesrotation(bladesrotation);
	}
}

void cycles_camera_set_focaldistance(ccl::Session* session_id, float focaldistance)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_focaldistance(focaldistance);
	}
}

void cycles_camera_set_shuttertime(ccl::Session* session_id, float shuttertime)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_shuttertime(shuttertime);
	}
}

void cycles_camera_set_fisheye_fov(ccl::Session* session_id, float fisheye_fov)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_fisheye_fov(fisheye_fov);
	}
}

void cycles_camera_set_fisheye_lens(ccl::Session* session_id, float fisheye_lens)
{
	ccl::Scene* sce = nullptr;
	if(scene_find(session_id, &sce)) {
		sce->camera->set_fisheye_lens(fisheye_lens);
	}
}

File: csycles/Camera.cs
/**
Copyright 2014-2024 Robert McNeel and Associates

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
**/

using System.Drawing;

namespace ccl
{
	/// <summary>
	/// Camera representation in a Cycles scene.
	/// </summary>
	public class Camera
	{
		/// <summary>
		/// Reference to the scene in which this camera is contained.
		/// </summary>
		internal Session Session { get; set; }

		/// <summary>
		/// Create a new camera representation for Session.
		/// </summary>
		/// <param name="scene"></param>
		internal Camera(Session session)
		{
			Session = session;
		}

		/// <summary>
		/// Set the camera size. This corresponds to the final pixel resolution of the
		/// image to be rendered.
		/// </summary>
		public Size Size
		{
			set
			{
				CSycles.camera_set_size(Session.Id, (uint)value.Width, (uint)value.Height);
			}
			get
			{
				return new Size((int)CSycles.camera_get_width(Session.Id), (int)CSycles.camera_get_height(Session.Id));
			}
		}

		/// <summary>
		/// Set the transformation matrix.
		/// </summary>
		public Transform Matrix
		{
			set
			{
				CSycles.camera_set_matrix(Session.Id, value);
			}
		}

		/// <summary>
		/// Set camera type.
		/// </summary>
		public CameraType Type
		{
			set
			{
				CSycles.camera_set_type(Session.Id, value);
			}
		}

		/// <summary>
		/// Set panorama type.
		/// </summary>
		public PanoramaType PanoramaType
		{
			set
			{
				CSycles.camera_set_panorama_type(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the Field of View.
		/// </summary>
		public float Fov
		{
			set
			{
				CSycles.camera_set_fov(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the sensor width.
		/// </summary>
		public float SensorWidth
		{
			set
			{
				CSycles.camera_set_sensor_width(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the sensor height.
		/// </summary>
		public float SensorHeight
		{
			set
			{
				CSycles.camera_set_sensor_height(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the near clip.
		/// </summary>
		public float NearClip
		{
			set
			{
				CSycles.camera_set_nearclip(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the far clip.
		/// </summary>
		public float FarClip
		{
			set
			{
				CSycles.camera_set_farclip(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the aperture size.
		/// </summary>
		public float ApertureSize
		{
			set
			{
				CSycles.camera_set_aperturesize(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the aperture ration.
		/// </summary>
		public float ApertureRatio
		{
			set
			{
				CSycles.camera_set_aperture_ratio(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the number of blades to use.
		/// </summary>
		public uint Blades
		{
			set
			{
				CSycles.camera_set_blades(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the blade rotation
		/// </summary>
		public float BladesRotation
		{
			set
			{
				CSycles.camera_set_bladesrotation(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the focal distance (for DoF)
		/// </summary>
		public float FocalDistance
		{
			set
			{
				CSycles.camera_set_focaldistance(Session.Id, value);
			}
		}

		/// <summary>
		/// Set the shutter time.
		/// </summary>
		public float ShutterTime
		{
			set
			{
				CSycles.camera_set_shuttertime(Session.Id, value);
			}
		}

		/// <summary>
		/// Set Field of View, in case fish eye projection is used.
		/// </summary>
		public float FishEyeFov
		{
			set
			{
				CSycles.camera_set_fisheye_fov(Session.Id, value);
			}
		}

		/// <summary>
		/// Set lens length, in case fish eye projection is used.
		/// </summary>
		public float FishEyeLens
		{
			set
			{
				CSycles.camera_set_fisheye_lens(Session.Id, value);
			}
		}

		/// <summary>
		/// Tag the camera for update.
		/// </summary>
		public void Update()
		{
			CSycles.camera_update(Session.Id);
		}

		/// <summary>
		/// Compute auto view plane.
		/// </summary>
		public void ComputeAutoViewPlane()
		{
			CSycles.camera_compute_auto_viewplane(Session.Id);
		}

		/// <summary>
		/// Set view plane
		/// </summary>
		/// <param name="left"></param>
		/// <param name="right"></param>
		/// <param name="top"></param>
		/// <param name="bottom"></param>
		public void SetViewPlane(float left, float right, float top, float bottom)
		{
			CSycles.camera_set_viewplane(Session.Id, left, right, top, bottom);
		}
	}
}
